pkg/activator/test/roundtripper.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/queue"
+)
+
+// FakeResponse is a response given by the FakeRoundTripper
+type FakeResponse struct {
+	Err  error
+	Code int
+	Body string
+}
+
+// FakeRoundTripper is a roundtripper emulator useful in testing
+type FakeRoundTripper struct {
+	T               *testing.T
+	ExpectHost      string
+	LockerCh        chan struct{}
+	ProbeResponses  []FakeResponse
+	RequestResponse *FakeResponse
+	mux             sync.Mutex
+}
+
+func defaultProbeResponse() *FakeResponse {
+	return &FakeResponse{
+		Err:  nil,
+		Code: 200,
+		Body: queue.Name,
+	}
+}
+
+func defaultRequestResponse() *FakeResponse {
+	return &FakeResponse{
+		Err:  nil,
+		Code: 200,
+		Body: "default response",
+	}
+}
+
+func response(fr *FakeResponse) (*http.Response, error) {
+	recorder := httptest.NewRecorder()
+	recorder.WriteHeader(fr.Code)
+	recorder.WriteString(fr.Body)
+	return recorder.Result(), nil
+}
+
+func popResponse(responses []FakeResponse, fallback *FakeResponse) (*FakeResponse, []FakeResponse) {
+	if responses == nil || len(responses) == 0 {
+		return fallback, responses
+	}
+	resp := &responses[0]
+	if len(responses) > 1 {
+		responses = responses[1:]
+	}
+	return resp, responses
+}
+
+// RT is a RoundTripperFunc
+func (rt *FakeRoundTripper) RT(req *http.Request) (*http.Response, error) {
+	var resp *FakeResponse
```

Comment:
Nit: I'd ditch the outer `resp` variable and return early on the probe path to make the general request path less indented.